File: frontend/src/models/Figure.ts
export enum FigureColor {
  WHITE = "white",
  BLACK = "black"
}

export interface Figure {
  id: string;
  name: string;
  color: FigureColor;
  position: [number, number];
}

export class Pawn implements Figure {
  id: string;
  name: string = "pawn";
  color: FigureColor;
  position: [number, number];

  constructor(id: string, color: FigureColor, position: [number, number]) {
      this.id = id;
      this.color = color;
      this.position = position;
  }
}

export class Rook implements Figure {
  id: string;
  name: string = "rook";
  color: FigureColor;
  position: [number, number];
  hasMoved: boolean = false;

  constructor(id: string, color: FigureColor, position: [number, number]) {
      this.id = id;
      this.color = color;
      this.position = position;
  }
}

export class Knight implements Figure {
  id: string;
  name: string = "knight";
  color: FigureColor;
  position: [number, number];

  constructor(id: string, color: FigureColor, position: [number, number]) {
      this.id = id;
      this.color = color;
      this.position = position;
  }
}

export class Bishop implements Figure {
  id: string;
  name: string = "bishop";
  color: FigureColor;
  position: [number, number];

  constructor(id: string, color: FigureColor, position: [number, number]) {
      this.id = id;
      this.color = color;
      this.position = position;
  }
}

export class Queen implements Figure {
  id: string;
  name: string = "queen";
  color: FigureColor;
  position: [number, number];

  constructor(id: string, color: FigureColor, position: [number, number]) {
      this.id = id;
      this.color = color;
      this.position = position;
  }
}

export class King implements Figure {
  id: string;
  name: string = "king";
  color: FigureColor;
  position: [number, number];
  hasMoved: boolean = false;

  constructor(id: string, color: FigureColor, position: [number, number]) {
      this.id = id;
      this.color = color;
      this.position = position;
  }
}

File: backend/models/figure.py
import uuid
from pydantic import BaseModel, Field
from enum import Enum

class FigureColor(str, Enum):
    WHITE = "white"
    BLACK = "black"

class Figure(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    name: str
    color: FigureColor
    position: tuple[int, int]
    
    @property
    def image_path(self) -> str:
        return f"/assets/{self.color}_{self.name}.png"

class Pawn(Figure):
    name: str = "pawn"

class Rook(Figure):
    name: str = "rook"
    has_moved: bool = False

class Knight(Figure):
    name: str = "knight"

class Bishop(Figure):
    name: str = "bishop"

class Queen(Figure):
    name: str = "queen"

class King(Figure):
    name: str = "king"
    has_moved: bool = False
